File: Where/Where/Src/IDMaker.m
//
//  IDMaker.m
//  Where
//
//  Created by Nero on 10/9/15.
//  Copyright (c) 2015 Nero. All rights reserved.
//

#import "IDMaker.h"

static IDMaker* _instance = nil;

@interface IDMaker () {
    int idPool;
    NSMutableArray *unUseIDs;
}

@end

@implementation IDMaker

+ (IDMaker*) sharedInstance {
    if (nil == _instance) {
        _instance = [IDMaker new];
    }
    return _instance;
}

- (id) init {
    self = [super init];
    if (nil != self) {
        idPool = 0;
        unUseIDs = [NSMutableArray new];
    }
    return self;
}

- (int) retainID {
    int result = -1;
    if ([unUseIDs count] != 0 && [unUseIDs isKindOfClass:[NSNumber class]] == YES ) {
        result = [unUseIDs.lastObject intValue];
        [unUseIDs removeLastObject];
    } else {
        result = ++idPool;
    }
    return result;
}

- (void) releaseID : (int) idNumber {
    if (idNumber <= idPool) {
        BOOL allowRelease = YES;
        for (NSNumber* number in unUseIDs) {
            if (idNumber == [number intValue]) {
                allowRelease = NO;
                break;
            }
        }
        if (YES == allowRelease) {
            [unUseIDs addObject:[[NSNumber alloc] initWithInt:idNumber]];
        }
    }
}

@end

File: Where?/Where?/Class/IDMaker.swift
//
//  IDMaker.swift
//  Where?
//
//  Created by Nero on 10/10/15.
//  Copyright (c) 2015 Nero. All rights reserved.
//

import Foundation

class IDMaker {
    //错误的id
    static let nops : Int = -1;
    
    private static var instance : IDMaker? = nil
    
    //当前生成器
    private var idIndex = 0
    
    //获取下一个可用id,当id益处时返回无效id
    private var nextID : Int {
        get {
            if (idIndex < Int.max) {
                return ++idIndex
            } else {
                return IDMaker.nops;
            }
        }
    }
    
    //空闲id池
    private var idPool = [Int]()
    
    class func sharedInstance()->IDMaker! {
        if (nil == instance) {
            instance = IDMaker();
        }
        return instance;
    }
    
    //获取一个没有被使用过的唯一id
    func retainID()->Int {
        if (idPool.count == 0) {
            return nextID;
        } else {
            let id = idPool.last!
            idPool.removeLast()
            return id
        }
    }
    
    //收回一个之前被占用了的id
    func releaseID(id : Int)->Bool {
        if (id < idIndex && ArraySearch(idPool, search: id)) {
            idPool.append(id)
            return true
        } else {
            return false
        }
    }
}